x/interchainquery/keeper/msg_server.go

```diff
@@ -28,105 +28,143 @@ func NewMsgServerImpl(keeper Keeper) types.MsgServer {
+		moduleCallbackHandler := k.callbacks[module]
 
-			merkleProof, err := commitmenttypes.ConvertProofs(msg.ProofOps)
+		if moduleCallbackHandler.Has(q.CallbackId) {
+			k.Logger(ctx).Info(fmt.Sprintf("ICQ Callback (%s) found for module (%s)", q.CallbackId, module))
+			// call the correct callback function
+			err := moduleCallbackHandler.Call(ctx, q.CallbackId, msg.Result, q)
 			if err != nil {
-				return nil, fmt.Errorf("error converting proofs")
+				k.Logger(ctx).Error(fmt.Sprintf("error in ICQ callback, error: %s, msg: %s, result: %v, type: %s, params: %v", err.Error(), msg.QueryId, msg.Result, q.QueryType, q.Request))
+				return err
 			}
+		} else {
+			k.Logger(ctx).Info(fmt.Sprintf("ICQ Callback not found for module (%s)", module))
+		}
+	}
+	return nil
+}
 
-			tmclientstate, ok := clientState.(*tmclienttypes.ClientState)
-			if !ok {
-				return nil, fmt.Errorf("error unmarshaling client state %v", clientState)
-			}
+// verify the query has not exceeded its ttl
+func (k Keeper) HasQueryExceededTtl(ctx sdk.Context, msg *types.MsgSubmitQueryResponse, q types.Query) (bool, error) {
+	k.Logger(ctx).Info(fmt.Sprintf("[ICQ Resp] query %s with ttl: %d, resp time: %d.", msg.QueryId, q.Ttl, ctx.BlockHeader().Time.UnixNano()))
+	curT, err := cast.ToUint64E(ctx.BlockTime().UnixNano())
+	if err != nil {
+		return false, err
+	}
 
-			if len(msg.Result) != 0 {
-				// if we got a non-nil response, verify inclusion proof.
-				if err := merkleProof.VerifyMembership(tmclientstate.ProofSpecs, consensusState.GetRoot(), path, msg.Result); err != nil {
-					return nil, fmt.Errorf("unable to verify proof: %s", err)
-				}
-				k.Logger(ctx).Info(fmt.Sprintf("Proof validated! module: %s, queryId %s", types.ModuleName, q.Id))
-
-			} else {
-				// if we got a nil response, verify non inclusion proof.
-				if err := merkleProof.VerifyNonMembership(tmclientstate.ProofSpecs, consensusState.GetRoot(), path); err != nil {
-					return nil, fmt.Errorf("unable to verify proof: %s", err)
-				}
-				k.Logger(ctx).Info(fmt.Sprintf("Non-inclusion Proof validated! module: %s, queryId %s", types.ModuleName, q.Id))
-			}
+	if q.Ttl < curT {
+		errMsg := fmt.Sprintf("[ICQ Resp] aborting query callback due to ttl expiry! ttl is %d, time now %d for query of type %s with id %s, on chain %s",
+			q.Ttl, ctx.BlockHeader().Time.UnixNano(), q.QueryType, q.ChainId, msg.QueryId)
+		k.DeleteQuery(ctx, msg.QueryId)
+		k.Logger(ctx).Error(errMsg)
+		return true, nil
+	}
+	return false, nil
+}
+
+func (k msgServer) SubmitQueryResponse(goCtx context.Context, msg *types.MsgSubmitQueryResponse) (*types.MsgSubmitQueryResponseResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	// 1. check if the response has an associated query stored on stride
+	q, found := k.GetQuery(ctx, msg.QueryId)
+	if found {
+		// 2. verify the query's ttl is unexpired
+		exceeded, err := k.HasQueryExceededTtl(ctx, msg, q)
```

Comment:
wdyt of pulling `k.DeleteQuery(ctx, msg.QueryId)` out and nesting it under `if exceeded {` below?

alternatively, we could rename `HasQueryExceededTtl` to `DeleteQueryIfExceededTtl`